ai2thor/controller.py

```diff
@@ -960,18 +968,27 @@ def step(self, action=None, **action_args):
         self.server.send(action)
         try:
             self.last_event = self.server.receive()
-        except UnityCrashException as e:
+        except UnityCrashException:
             self.server.stop()
             self.server = None
             # we don't need to pass port or host, since this Exception
             # is only thrown from the FifoServer, start_unity is also
             # not passed since Unity would have to have been started
             # for this to be thrown
-            message = "Restarting unity due to crash: %s" % e
+            message = (
+                f"Restarting unity due to crash when when running action {action}"
+                f" in scene {self.last_event.metadata['sceneName']}:\n{traceback.format_exc()}"
+            )
             warnings.warn(message)
             self.start(width=self.width, height=self.height, x_display=self.x_display)
             self.reset()
             raise RestartError(message)
+        except Exception as e:
+            self.server.stop()
+            raise (TimeoutError if isinstance(e, TimeoutError) else RuntimeError)(
```

Comment:
Would it be more clear if we had an `except` branch for `TimeoutError`?

Reply:
I agree regarding clarity but I'd prefer to have a single branch here as, otherwise, we would be duplicating the same `self.server.stop()` etc code twice.